andela/activo-api
# #157556300 Create authentication decorator for backend APIs

#### What does this PR do?
Create authentication decorator for backend APIs
#### Description of Task to be completed?
A `GET` to `/` responding with ‘Unauthorised’ with a status code of 401 when invocation is made without the token.  
#### How should this be manually tested?
- Pull from this branch
- run ```pip install -r requirements.txt```
- Set ```FLASK_ENV=development```, ```JWT_PUBLIC_KEY```  in the ```.env``` file
- run ```python manage.py runserver```
- open postman hit the `GET` `/` route with the parameters provided in the screenshot below

                                                OR

- open your terminal and run these commands:
- `curl http://127.0.0.1:5000/` 
   Response
 {
        "message": "Bad request. Header does not contain an authorization token.",
        "status": "Failure", 
        "status_code": 400
}

- `curl --header "Authorization: bad token" http://127.0.0.1:5000/ `
   Response
{
   "message": "Bad request. The token should begin with the word 'Bearer'.",
    "status": "Failure",
     "status_code":400
}

- `curl --header "Authorization: Bearer bad token" http://127.0.0.1:5000/ ` 
  Response
{
  "message": "Authorization failed due to an Invalid token.",
  "status": "Failure", 
  "status_code":401
}

- to get the `500` error remove the ```JWT_PUBLIC_KEY``` from the ```.env``` file and run`curl --header "Authorization: Bearer bad token" http://127.0.0.1:5000/ ` 
  Response
{
  "message": "Authorization failed. Please contact support.",
  "status": "Failure", 
  "status_code":500
}

- `curl --header "Authorization: Bearer good token" http://127.0.0.1:5000/ ` 
  Response
{
  "message": "Welcome to the AM api",
}

#### Any background context you want to provide?
I created a base validation class for returning validation errors when errors are encountered
#### What are the relevant pivotal tracker stories?
#157556300
#### Screenshots (if appropriate)
<img width="1440" alt="screen shot 2018-05-16 at 12 49 19 pm" src="https://user-images.githubusercontent.com/25517388/40116529-d0430084-590b-11e8-8a96-c8afd82406d0.png">
<img width="1440" alt="screen shot 2018-05-16 at 1 21 24 pm" src="https://user-images.githubusercontent.com/25517388/40116611-12201460-590c-11e8-9167-87aa3b4313b5.png">
<img width="1440" alt="screen shot 2018-05-16 at 1 22 52 pm" src="https://user-images.githubusercontent.com/25517388/40116686-5247501c-590c-11e8-9842-0ec6231e1a43.png">

#### Questions:
N/A
